Support wpt run safari (Safari 11 and Technology Preview)

gsnedders opened the pull request:
Originally posted as https://github.com/w3c/wptrunner/pull/225 by @jgraham on 18 Jan 2017, 13:06 UTC:

> 
> 
> <!-- Reviewable:start -->
> This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/w3c/wptrunner/225)
> <!-- Reviewable:end -->

jgraham committed:
Add Safari support.

Add support for Safari >= 10 using the built-in WebDriver implementation.
Due to Safari security requirements, this currently requires manual configuration
to enable remote automation and disable the popup blocker.

wpt-issue-mover commented:
Originally posted as https://github.com/w3c/wptrunner/pull/225#issuecomment-275313777 by @gsnedders on 26 Jan 2017, 05:59 UTC:

> I feel like we're getting more and more duplicated code across the Selenium-using executors. :(

wpt-issue-mover commented:
Originally posted as https://github.com/w3c/wptrunner/pull/225#issuecomment-276103333 by @andreastt on 30 Jan 2017, 16:04 UTC:

> Also the inheritance surrounding overriding `WebDriverServer`’s ctor isn’t great, but this looks like a pretty low-impact change. I’m happy with it.

wpt-issue-mover commented:
Originally posted as https://github.com/w3c/wptrunner/pull/225#issuecomment-276103648 by @andreastt on 30 Jan 2017, 16:05 UTC:

> I meant “low impact” as in “not dangerous” or controversial; not that having Safari support for WPT is. On the contrary: having that is excellent!

gsnedders commented on tools/wptrunner/wptrunner/browsers/safari.py:
```diff
@@ -0,0 +1,79 @@
+# This Source Code Form is subject to the terms of the Mozilla Public
+# License, v. 2.0. If a copy of the MPL was not distributed with this file,
+# You can obtain one at http://mozilla.org/MPL/2.0/.
```
This presumably is out of date now?

foolip replied:
@jgraham, OK to just delete this?

gsnedders commented on tools/wptrunner/wptrunner/browsers/safari.py:
```diff
@@ -0,0 +1,79 @@
+        return self.server.pid
+
+    def is_alive(self):
+        # TODO(ato): This only indicates the driver is alive,
```
@andreastt ?

andreastt replied:
I think this is fine. This is the [same thing we do for Chrome](https://searchfox.org/mozilla-central/rev/03877052c151a8f062eea177f684a2743cd7b1d5/testing/web-platform/tests/tools/wptrunner/wptrunner/browsers/chrome.py#95) and we could probably just remove this comment.

foolip replied:
It's sprinkled all over, I suggest leaving it here or removing it everywhere.

w3c-bots commented:
# Build PASSED

Started: 2018-01-10 16:25:33
Finished: 2018-01-10 16:48:44





View more information about this build on:

- [WPT PR Status](http://pulls.web-platform-tests.org/build/22587)
- [TravisCI](https://travis-ci.org/w3c/web-platform-tests/builds/327289315)

andreastt reviewed (approved):
Thanks @gsnedders for resurrecting this.

If this works in real life and passes the stability tests on Travis I’m OK with this change landing as it is.

gsnedders commented:
@andreastt I literally haven't tested it (and we might run into https://github.com/w3c/web-platform-tests/issues/6998 here too), I just rebased it onto the mono repo. Should actually check it works next. :)

jgraham reviewed (changes requested):
I think some changes are needed here.

foolip commented:
I've pushed a commit which gets this closer to working. With `OBJC_DISABLE_INITIALIZE_FORK_SAFETY=YES ./wpt run --no-manifest-update --log-mach-level debug safari dom/historical.html` I get this:
```
 0:04.33 INFO Using 1 client processes
 0:04.38 SUITE_START: web-platform-test - running 1 tests
 0:04.38 INFO Running reftest tests
 0:04.39 INFO Starting http server on web-platform.test:8000
 0:04.39 INFO Starting http server on web-platform.test:8001
 0:04.40 INFO Starting http server on web-platform.test:8443
 0:04.43 DEBUG Using 1 processes
 0:04.43 INFO No reftest tests to run
 0:04.43 DEBUG Stop flag set in ManagerGroup
 0:04.43 INFO Running wdspec tests
 0:04.44 DEBUG Using 1 processes
 0:04.44 INFO No wdspec tests to run
 0:04.44 DEBUG Stop flag set in ManagerGroup
 0:04.44 INFO Running testharness tests
 0:04.44 DEBUG Using 1 processes
 0:04.44 DEBUG Dispatch start_init
 0:04.44 DEBUG new state: initializing_browser
 0:04.44 DEBUG Dispatch init
 0:04.44 DEBUG Init called, starting browser and runner
 0:04.44 DEBUG Starting browser with settings {}
 0:04.45 DEBUG Waiting for server to become accessible: http://127.0.0.1:4446/
 0:04.95 INFO Starting runner
 0:04.95 DEBUG Test runner started
 0:04.96 DEBUG Got command: u'log'
 0:04.96 DEBUG Executor setup
 0:34.44 DEBUG init_failed called from timer
 0:34.44 DEBUG Got command: u'init_failed'
 0:34.47 DEBUG ensure_runner_stopped
 0:34.47 DEBUG waiting for runner process to end
 0:34.47 DEBUG After join
 0:34.47 DEBUG Testrunner exited with code -11
 0:34.47 DEBUG TestManager cleanup
 0:34.47 DEBUG new state: initializing_browser
 0:34.47 DEBUG Dispatch init
 0:34.47 DEBUG Init called, starting browser and runner
 0:34.47 DEBUG Starting browser with settings {}
 0:34.48 DEBUG Waiting for server to become accessible: http://127.0.0.1:4446/
 0:34.98 INFO Starting runner
 0:34.98 DEBUG Test runner started
 0:34.99 DEBUG Got command: u'log'
 0:34.99 DEBUG Executor setup
 0:35.00 DEBUG Got command: u'log'
 0:35.00 DEBUG Connecting to Selenium on URL: http://127.0.0.1:4446/
 0:35.45 DEBUG Got command: u'log'
 0:35.45 DEBUG Selenium session started
 0:35.45 DEBUG Got command: u'log'
 0:35.45 DEBUG Loading http://web-platform.test:8000/testharness_runner.html
 0:35.69 DEBUG Got command: 'init_succeeded'
 0:35.69 DEBUG new state: running
 0:35.69 DEBUG Dispatch run_test
 0:35.69 TEST_START: /dom/historical.html
 0:35.69 DEBUG Got command: u'log'
 0:35.69 DEBUG Executor setup done
 0:35.73 DEBUG Got command: u'log'
 0:35.73 DEBUG unable to find test window
 0:35.73 DEBUG Got command: 'test_ended'
 0:35.73 DEBUG Unexpected count in this thread 1
 0:35.73 TEST_END: ERROR, expected OK
unable to find test window
 0:35.73 INFO Pausing until the browser exits
```

The "DEBUG Executor setup" step takes a long time, and finally the rest happens. I only see one Safari window opened, so I wonder if it's a popup blocker issue.

foolip commented:
Results are similar with `--webdriver-binary "/Applications/Safari Technology Preview.app/Contents/MacOS/safaridriver"`, but it's promising how similar it is, this approach will probably work with both.

@burg @youennf FYI that we're playing with this, in case you want to swoop in and rescue us :)

foolip commented:
Sweet, disabling the popup blocker in Safari TP preferences and it actually works. Kind of. After the test is run we get stuck at "INFO Pausing until the browser exits". Stopping the automation session makes the single test pass. What I saw:
![screen shot 2018-02-23 at 9 22 43 pm](https://user-images.githubusercontent.com/498917/36598729-bb48b834-18df-11e8-99c6-b88d364b1bad.png)
That's all I have time for now, no promises that I'll come back and finish this work.

foolip commented:
@jgraham has informed me on IRC that "the default behaviour if you run a single test is to wait for you to close the window on the basis that if you run a single test you are probably debugging that test specifically" and "you can pass --pause[-after-test] to require human intervention after each test or --no-pause[-after-test] to never require it, and if you do neither the default is to require it when only a single test is run".

So this PR actually works, and I declare it ready for review. In order to actually make it work, one has to:
* run `safaridriver --enable`
* disable the popup blocking the the Safari settings
* use `OBJC_DISABLE_INITIALIZE_FORK_SAFETY=YES`

Not sure how to deal with that, suggestions welcome.

foolip commented:
This is also totally missing tests. @jgraham @gsnedders, are any of the corresponding bits for any other browser tested, and what kinds of test would be useful?

foolip commented:
A good and a bad finding:
* `--install-fonts` just works already and /infrastructure/assumptions/ahem.html passes
* the HTTPS cert isn't trusted, breaking all (?) .https. tests

foolip commented on tools/wpt/browser.py:
```diff
@@ -431,6 +431,28 @@ def version(self):
+class Safari(Browser):
+    """Safari-specific interface.
+
+    Includes installation, webdriver installation, and wptrunner setup methods.
```
This is copypasta and just wrong, but so are most of the others. Maybe update it in a separate PR before this one.

foolip commented:
On the need for `OBJC_DISABLE_INITIALIZE_FORK_SAFETY=YES`, it unfortunately seems like that has to be set in the environment *before* invoking even the outermost `wpt` python script, setting it in the first line of the script doesn't work.

foolip commented:
There's a relevant thread on webkit-dev, and in particular this reply from @burg:
https://lists.webkit.org/pipermail/webkit-dev/2018-February/029866.html

@burg, I take it this means that the WebDriver tests (`--test-types wdspec`) aren't expected to pass in SafariDriver at all? I did manage to run the directory with my machine catching fire, and not everything failed, but maybe it's still a pointless exercise? If so we should try to figure out how to just skip those test until SafariDriver supports it.

foolip committed:
Finish support for Safari 11 and Safari Technology Preview

Whether this works with Safari 10 (parent commit) has not been tested.

Also improve the debug logging around WebDriver start and document the
manual setup required to make this work.

foolip commented:
I've added the documentation and rebased into two commits, the shape that I think the PR should be merged in. Even more ready for review :)

jgraham commented:
Reviewed 1 of 2 files at r2, 2 of 5 files at r3, 1 of 1 files at r4, 6 of 6 files at r5.
Review status: all files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[tools/wpt/run.py, line 332 at r5](https://reviewable.io:443/reviews/w3c/web-platform-tests/8979#-L6RvLiZ_4q5ORVVGR3T:-L6RvLiZ_4q5ORVVGR3U:bwg2bfw) ([raw file](https://github.com/w3c/web-platform-tests/blob/3f5ec52346d84e26ee3045e94f3605636f45c36e/tools/wpt/run.py#L332)):*
> ```Python
>                 raise WptrunError("Unable to locate safaridriver binary")
> 
>             kwargs["webdriver_binary"] = self.browser.find_webdriver()
> ```

I assume you meant `kwargs["webdriver_binary"] = webdriver_binary`

---

*[tools/wptrunner/requirements_safari.txt, line 1 at r5](https://reviewable.io:443/reviews/w3c/web-platform-tests/8979#-L6RvUT3OG5OegVkSxZX:-L6RvUT3OG5OegVkSxZY:b-4fc3t1) ([raw file](https://github.com/w3c/web-platform-tests/blob/3f5ec52346d84e26ee3045e94f3605636f45c36e/tools/wptrunner/requirements_safari.txt#L1)):*
> ```
> mozprocess == 0.26
> ```

You probably want to add this file to the .pyup.yml file in the root.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/w3c/web-platform-tests/8979)*
<!-- Sent from Reviewable.io -->

jgraham reviewed (approved):
There's a couple of trivial changes to make, but I'm happy for you to land without further review.

gsnedders commented on docs/_running-tests/safari.md:
```diff
@@ -0,0 +1,23 @@
+---
+To run Safari on macOS, some manual setup is required:
+
+  * Disable "Block pop-up windows" in the security preferences.
```
This should be doable with `defaults`. We should be able to set this when we start the browser and reset it to its previous value after running.

gsnedders commented on docs/_running-tests/safari.md:
```diff
@@ -0,0 +1,23 @@
+
+  * Disable "Block pop-up windows" in the security preferences.
+
+  * Ensure that "Allow Remote Automation" is enabled in the "Develop" menu.
```
`safaridriver --enable` should work. (Throws admin prompt to authentication unless run as root, though.)

burg replied:
Yes. Also, --enable is first supported in Safari 11.

gsnedders commented on docs/_running-tests/safari.md:
```diff
@@ -0,0 +1,23 @@
+
+  * Ensure that "Allow Remote Automation" is enabled in the "Develop" menu.
+
+  * Import the web-platform.test certificate and root certificate authority in
```
Should be doable with `security`. (Again throws admin prompt unless root.)

gsnedders commented on docs/_running-tests/safari.md:
```diff
@@ -0,0 +1,23 @@
+    manually mark the root as trusted.
+
+  * Set `OBJC_DISABLE_INITIALIZE_FORK_SAFETY=YES` in your environment. This is a
+    workaround for https://github.com/w3c/web-platform-tests/issues/9007.
```
Please make this URL a link.

burg commented on docs/_running-tests/safari.md:
```diff
@@ -0,0 +1,24 @@
+    `tools/certs/` into the keychain of the user that will run the tests, and
+    manually mark the root as trusted.
+
+  * Set `OBJC_DISABLE_INITIALIZE_FORK_SAFETY=YES` in your environment. This is a
```
Is this specific to Safari support, or is this a larger wpt issue? safaridriver itself doesn't fork.

sideshowbarker replied:
```diff
+  * Set `OBJC_DISABLE_INITIALIZE_FORK_SAFETY=YES` in your environment. This is a
```
> Is this specific to Safari support, or is this a larger wpt issue?

It’s a larger wpt issue, not specific to Safari support. It seems to be an issue/bug in Python2 around how it handles multiprocessing on MacOS. See https://github.com/w3c/web-platform-tests/issues/9007 and https://bugs.python.org/issue30837

gsnedders replied:
Larger `wptrunner` issue (because it does fork).

foolip replied:
Yeah, if we had docs for Chrome on macOS it would have to say the same thing. A fix for that issue would be good, but until then documenting it helps.s

foolip committed:
tweak documentation

foolip committed:
Document the command line alternatives for `wpt run safari` setup

There will still be authenitication dialogs, so it's still manual.

foolip committed:
find_binary

foolip committed:
drop SafariDriverWdspecExecutor

foolip commented:
OK, I managed to find commands for all the setup, although it's still manual. And I dropped SafariDriverWdspecExecutor which means the webdriver tests will be skipped.

I no longer plan to merge this as multiple commits, since the first wouldn't work. Will squash and merge.
